### PR TITLE
AP_DDS: GeoPose Publisher

### LINKED ISSUE #23534: AP_DDS: Support publishing GeoPoseStamped
I always viewed `NavSatFix` as rawish sensor data more than an actual position output.
Similar to the way for orientation we use a `Pose` message with a `Quaternion` field, and not, say, a `MagneticField` (https://docs.ros2.org/foxy/api/sensor_msgs/msg/MagneticField.html) message.

As it's appropriate here, I'll mention the option of using the `geographic_msgs` `GeoPoseStamped` (https://github.com/ros-geographic-info/geographic_info/blob/ros2/geographic_msgs/msg/GeoPoseStamped.msg) type.

_Originally posted by @russkel in https://github.com/ArduPilot/ardupilot/issues/23277#issuecomment-1499895214_

This ticket is to add GeoPoseStamped support, which is the fused GPS position of the vehicle (from AHRS).

### PR DESCRIPTION
**Pull request by pedro-fuoco**
Resolves #23534, here is the PR working on SITL:
Screenshot from 2023-04-18 15-37-54
Uses the same orientation logic as #23480 since both follow ENU

### CODE REVIEW

File: libraries/AP_DDS/AP_DDS_Client.cpp [hunk 1]
```diff
@@ -234,6 +235,46 @@ void AP_DDS_Client::update_topic(sensor_msgs_msg_BatteryState& msg, const uint8_
...
+        msg.pose.orientation.z = -orientation[3];
+        msg.pose.orientation.w = orientation[0];
+    }
+
+    Location loc;
+    if (ahrs.get_location(loc))
+    {
+        msg.pose.position.latitude = loc.lat * 1E-7;
+        msg.pose.position.longitude = loc.lng * 1E-7;
+        msg.pose.position.altitude = loc.alt/100.0; // Transform from cm to m
```
Comment (Ryanf55) [inline libraries/AP_DDS/AP_DDS_Client.cpp:274]: Add spaces around the `/` operator please.

File: libraries/AP_DDS/AP_DDS_Client.h [hunk 2]
```diff
@@ -56,13 +58,16 @@ class AP_DDS_Client
...
     static void update_topic(sensor_msgs_msg_BatteryState& msg, const uint8_t instance);
+    static void update_topic(geographic_msgs_msg_GeoPoseStamped& msg);
 
     // The last ms timestamp AP_DDS wrote a Time message
     uint64_t last_time_time_ms;
     // The last ms timestamp AP_DDS wrote a NavSatFix message
     uint64_t last_nav_sat_fix_time_ms;
     // The last ms timestamp AP_DDS wrote a BatteryState message
     uint64_t last_battery_state_time_ms;
+    // The last ms timestamp AP_DDS wrote a Geo Pose message
```
Comment (Ryanf55) [inline libraries/AP_DDS/AP_DDS_Client.h:69]: GeoPose can be one word.

File: libraries/AP_DDS/AP_DDS_Client.h [hunk 3]
```diff
@@ -88,6 +93,8 @@ class AP_DDS_Client
     void write_static_transforms();
     //! @brief Serialize the current nav_sat_fix state and publish it to the IO stream(s)
     void write_battery_state_topic();
+    //! @brief Serialize the current geo pose and publish to the IO stream(s)
```
Comment (Ryanf55) [inline libraries/AP_DDS/AP_DDS_Client.h:96]: `geo_pose` with underscore

File: libraries/AP_DDS/README.md [hunk 4]
```diff
@@ -103,6 +103,11 @@ Follow the steps to use the microROS Agent
 
   - https://docs.ros.org/en/humble/Installation/Ubuntu-Install-Debians.html
 
+- Install geographic_msgs
+  ```bash
+  sudo apt install ros-humble-geographic-msgs
```
Comment (Ryanf55) [inline libraries/AP_DDS/README.md:108]: Later, once the testing PR that adds colcon support to this repo is added, dependencies can be added there to the package.xml. For now, this is fine.

**Review by Ryanf55 — CHANGES_REQUESTED** [review]

File: libraries/AP_DDS/README.md (hunk 4 above)
Comment (pedro-fuoco) [reply libraries/AP_DDS/README.md:108]: Agreed, we can just rely on rosdep when package.xml is ready

**Comment by pedro-fuoco** [comment]
Updated the orientation to follow what was agreed on #23480. Code comments were updates accordingly.

**Review by Ryanf55 — APPROVED** [review]
LGTM, I rebased and updated the code style and tested it on UDP. 
```bash
./Tools/CodeStyle/ardupilot-astyle.sh --options=Tools/CodeStyle/astylerc libraries/AP_DDS/*.cpp libraries/AP_DDS/*.h
```

Make sure your commits have a blank line between the first line and any further details.
```
AP_DDS: Add geographic_msgs 
* Add GeoPoint.idl
* Add GeoPose.idl
* Add GeoPoseStamped.idl
```
I squashed all the commits with the formatting. Approved

**Comment by pedro-fuoco** [comment]
@Ryanf55 Thanks a lot for the insights, I will surely keep it all in mind for my future contributions!

**Comment by russkel** [comment]
Lovely work.

**Pull request merged**